Kernel/Storage/ATA/GenericIDE/Channel.cpp

```diff
@@ -0,0 +1,316 @@
+    return adopt_ref(*new IDEChannel(controller, irq, io_group, type, move(ata_identify_data_buffer)));
+}
+
+StringView IDEChannel::channel_type_string() const
+{
+    if (m_channel_type == ChannelType::Primary)
+        return "Primary"sv;
+    return "Secondary"sv;
+}
+
+bool IDEChannel::wait_until_not_busy(bool slave, size_t milliseconds_timeout)
+{
+    IO::delay(20);
+    m_io_group.io_base().offset(ATA_REG_HDDEVSEL).out<u8>(0xA0 | (slave << 4)); // First, we need to select the drive itself
+    IO::delay(20);
+    size_t time_elapsed = 0;
+    while (m_io_group.control_base().in<u8>() & ATA_SR_BSY && time_elapsed <= milliseconds_timeout) {
+        IO::delay(1000);
+        time_elapsed++;
+    }
+    return time_elapsed <= milliseconds_timeout;
+}
+
+ErrorOr<void> IDEChannel::port_phy_reset()
+{
+    MutexLocker locker(m_lock);
+    SpinlockLocker hard_locker(m_hard_lock);
+    // reset the channel
+    u8 device_control = m_io_group.control_base().in<u8>();
+    // Wait 30 milliseconds
+    IO::delay(30000);
+    m_io_group.control_base().out<u8>(device_control | (1 << 2));
+    // Wait 30 milliseconds
+    IO::delay(30000);
+    m_io_group.control_base().out<u8>(device_control);
+    // Wait up to 30 seconds before failing
+    if (!wait_until_not_busy(false, 30000)) {
+        dbgln("IDEChannel: reset failed, busy flag on master stuck");
+        return Error::from_errno(EBUSY);
+    }
+    // Wait up to 30 seconds before failing
+    if (!wait_until_not_busy(true, 30000)) {
+        dbgln("IDEChannel: reset failed, busy flag on slave stuck");
+        return Error::from_errno(EBUSY);
+    }
+    return {};
+}
+
+UNMAP_AFTER_INIT ErrorOr<void> IDEChannel::allocate_resources(Badge<IDEController>, bool force_pio)
+{
+    dbgln_if(PATA_DEBUG, "IDEChannel: {} IO base: {}", channel_type_string(), m_io_group.io_base());
+    dbgln_if(PATA_DEBUG, "IDEChannel: {} control base: {}", channel_type_string(), m_io_group.control_base());
+    if (m_io_group.bus_master_base().has_value())
+        dbgln_if(PATA_DEBUG, "IDEChannel: {} bus master base: {}", channel_type_string(), m_io_group.bus_master_base().value());
+    else
+        dbgln_if(PATA_DEBUG, "IDEChannel: {} bus master base disabled", channel_type_string());
+
+    if (!force_pio) {
+        m_dma_enabled = true;
+        VERIFY(m_io_group.bus_master_base().has_value());
+        // Let's try to set up DMA transfers.
+        m_prdt_page = MM.allocate_supervisor_physical_page();
+        m_dma_buffer_page = MM.allocate_supervisor_physical_page();
+        if (m_dma_buffer_page.is_null() || m_prdt_page.is_null())
+            TODO();
+        {
+            auto region_or_error = MM.allocate_kernel_region(m_prdt_page->paddr(), PAGE_SIZE, "IDE PRDT", Memory::Region::Access::ReadWrite);
+            if (region_or_error.is_error())
+                TODO();
```

Comment:
As we have ErrorOr return type, we should change these to `TRY` return the error

Reply:
It's merely about merging both files together, so I prefer to do that in another commit (or maybe even in another PR).